src/shared/main-func.h

```diff
@@ -21,22 +23,49 @@
                 r = impl;                                               \
                 if (r < 0)                                              \
                         (void) sd_notifyf(0, "ERRNO=%i", -r);           \
-                (void) sd_notifyf(0, "EXIT_STATUS=%i", ret);            \
+                (void) sd_notifyf(0, "EXIT_STATUS=%i",                  \
+                                  result_to_exit_status(r));            \
                 ask_password_agent_close();                             \
                 polkit_agent_close();                                   \
                 pager_close();                                          \
                 mac_selinux_finish();                                   \
                 static_destruct();                                      \
-                return ret;                                             \
+                hashmap_trim_pools();                                   \
+                return result_to_return_value(r);                       \
         }
 
+static inline int fail_on_negative(int result) {
+        return result < 0 ? EXIT_FAILURE : EXIT_SUCCESS;
+}
+
 /* Negative return values from impl are mapped to EXIT_FAILURE, and
  * everything else means success! */
 #define DEFINE_MAIN_FUNCTION(impl)                                      \
-        _DEFINE_MAIN_FUNCTION(,impl(argc, argv), r < 0 ? EXIT_FAILURE : EXIT_SUCCESS)
+        _DEFINE_MAIN_FUNCTION(,impl(argc, argv), fail_on_negative, fail_on_negative)
+
+static inline int fail_on_nonzero(int result) {
+        return result < 0 ? EXIT_FAILURE : result;
+}
 
 /* Zero is mapped to EXIT_SUCCESS, negative values are mapped to EXIT_FAILURE,
  * and positive values are propagated.
  * Note: "true" means failure! */
 #define DEFINE_MAIN_FUNCTION_WITH_POSITIVE_FAILURE(impl)                \
-        _DEFINE_MAIN_FUNCTION(,impl(argc, argv), r < 0 ? EXIT_FAILURE : r)
+        _DEFINE_MAIN_FUNCTION(,impl(argc, argv), fail_on_nonzero, fail_on_nonzero)
+
+static inline int raise_or_exit(int ret) {
```

Comment:
this doesn't exit() like the name suggests.

maybe call this raise_or_exit_status() or so?